pkg/encryption/path.go

```diff
@@ -294,3 +302,73 @@ func decryptPathComponent(comp string, cipher storj.CipherSuite, key *storj.Key)
 
 	return string(decrypted), nil
 }
+
+// encodeSegment encodes segment according to specific rules
+// The empty path component is encoded as `\x01`
+// Any other path component is encoded as `\x02 + escape(component)`
+//
+// `\x2e` escapes to `\x2e\x01`
+// `\x2f` escapes to `\x2e\x02`
+// `\xfe` escapes to `\xfe\x01`
+// `\xff` escapes to `\xfe\x02`
+// `\x00` escapes to `\x01\x01`
+// `\x01` escapes to `\x01\x02
+// for more details see docs/design/path-component-encoding.md
+func encodeSegment(segment []byte) []byte {
+	if len(segment) == 0 {
+		return emptyComponent
+	}
+
+	result := make([]byte, 0, len(segment)*2+1)
+	result = append(result, notEmptyComponentPrefix)
+	for i := 0; i < len(segment); i++ {
+		switch segment[i] {
+		case escape1:
+			result = append(result, []byte{escape1, 1}...)
+		case escape1 + 1:
+			result = append(result, []byte{escape1, 2}...)
+		case escape2:
+			result = append(result, []byte{escape2, 1}...)
+		case escape2 + 1:
+			result = append(result, []byte{escape2, 2}...)
+		case escape3 - 1:
+			result = append(result, []byte{escape3, 1}...)
+		case escape3:
+			result = append(result, []byte{escape3, 2}...)
+		default:
+			result = append(result, segment[i])
+		}
+	}
+	return result
+}
+
+func decodeSegment(segment []byte) ([]byte, error) {
+	if len(segment) == 0 {
+		return segment, nil
+	}
+	if segment[0] == emptyComponent[0] {
+		return []byte{}, nil
+	}
+
+	if segment[0] != notEmptyComponentPrefix {
+		return []byte{}, errs.New("invalid encoded segment prefix")
+	}
+
+	currentIndex := 0
+	for i := 1; i < len(segment); i++ {
+		switch {
+		case i == len(segment)-1:
+			segment[currentIndex] = segment[i]
+		case segment[i] == escape1 || segment[i] == escape2:
+			segment[currentIndex] = segment[i] + segment[i+1] - 1
+			i++
+		case segment[i] == escape3:
+			segment[currentIndex] = segment[i+1] - 1
+			i++
+		default:
+			segment[currentIndex] = segment[i]
+		}
+		currentIndex++
+	}
```

Comment:
I think we need to be more strict in what we disallow, otherwise we can have distinct strings that map to the same value. For example, `\x02\x01\x23` and `\x02\x22` both map to `\x22`. This could possibly cause some havoc if a malicious client uploaded to the same path in multiple ways.

Can we enforce for some segment:

1. The last byte is not in `{escape1, escape2, escape3}`
2. Any byte after an escape character is `\x01` or `\x02`
3. It does not contain any characters in `{\x00, \xff, \x2f}`
4. It is non-empty
5. It begins with a character in `{\x01, \x02}`

I believe that's sufficient to enforce uniqueness, and `decodeSegment` can just call that function. Additionally, can we export a function that the satellite can call to check that a full encoded path is valid? That function would iterate over the segments and check that each segment is valid.

Reply:
Good idea, only thing I'm not sure if we should incorporate those rules as a separate function and use it in `decodeSegment` or put those rules directly into `decodeSegment` (might be faster). 

I think segment validation on satellite side is a good candidate for separate PR.

Reply:
I pushed method for encoded segment validation. I feel it can be more optimal but I would prefer to update this later.